### ISSUE TITLE
Feature request: Cursor-based pagination

### ISSUE THREAD

**Issue by chris-rock** [facebookincubator/ent#215]
The tool already includes support for pagination, but its limited to offset pagination which is known to be less optimal for fast-changing datasets. Support for cursor-based pagination would be awesome.

**Comment by a8m** [comment]
Hey @chris-rock, 
We actually discussed about this internally a few months and kinda agreed that we want to add a cursor-based API.

I think we'll address this after [v1](https://github.com/facebookincubator/ent/issues/46) is finished.

**Comment by chris-rock** [comment]
That sounds awesome. Thank you @a8m

**Comment by chris-rock** [comment]
@a8m I implemented [keyset pagination](https://use-the-index-luke.com/no-offset) with ent by using the following code:

```
func CompositeWhere(operator string, columns []string, args ...interface{}) func(*sql.Selector) {
	return func(s *sql.Selector) {
		total := s.Total()

		b := &sql.Predicate{}
		b.SetDialect(s.Dialect())
		b.Nested(func(nb *sql.Builder) {
			nb.IdentComma(columns...)
		})
		b.WriteString(operator)
		b.WriteString("(")
		b.Args(args...)
		b.WriteString(")")

		s.Where(b)
		s.SetTotal(total + b.Total())
	}
}

func CursorAsc(columns []string, args ...interface{}) func(s *sql.Selector) {
	const operator = " > "
	return CompositeWhere(operator, columns, args...)
}

func CursorDesc(columns []string, args ...interface{}) func(s *sql.Selector) {
	const operator = " < "
	return CompositeWhere(operator, columns, args...)
}
```
The code can be used then as follows:
```
// fetch cursor
cursorUser, err = client.Advisory.Query().Where(user.SubjectEQ(q.Page.Cursor)).Only(ctx)

// set fields for order and determine the query for the cursor
fieldOrder = client.Desc(user.FieldName)
idOrder = client.Desc(user.FieldSubject)
cursor = CursorDesc(Columns(user.FieldName, user.FieldId), cursorUser.Name, cursorUser.Subject)

data, err := client.Advisory.Query().
  Where(cursor).
  Order(fieldOrder, idOrder).
  Limit(pageSize).
  All(ctx)
```

I am happy to contribute if you think this is useful for this project.

**Comment by a8m** [comment]
Wonderful! I'm going to give it a look this weekend!
Thanks :)

**Comment by sohaha** [comment]
Are there any new developments?

**Comment by a8m** [comment]
> Are there any new developments?

No, we're currently focused only designing and developing the hooks and the privacy (authorization) layers.

**Comment by a8m** [comment]
Hey, we just open-sourced some of our GraphQL integration, and there's a pagination support there. It's currently based on the _Relay Cursor Connection_ spec, but it may be added to the core framework in the future. See [entgo.io/docs/graphql](https://entgo.io/docs/graphql/#pagination) (note these docs are WIP at the moment).

It's actually uses your work @chris-rock from #264 😉

**Comment by chris-rock** [comment]
Thank is awesome @a8m I am happy to see this being extended. I am also glad to see the integration with gqlgen. I hope I have some time soon to try this out.

**Comment by dilipkk-foyernet** [comment]
@a8m is there a way, we can get the number of total results to show the number of pages or results count in the UI?

**Comment by a8m** [comment]
> @a8m is there a way, we can get the number of total results to show the number of pages or results count in the UI?

The GraphQL integration supports this. Please follow the guide in the website and check the [todo app example](https://github.com/facebookincubator/ent-contrib/blob/33a6b7bbb091a6fdc471917bb582da5e09adc506/entgql/internal/todo/todo.graphql#L39) in ent-contrib.

**Comment by BrysonUrieDakland** [comment]
@a8m Are there any plans to add support for this outside of the GraphQL API?
